### DISCUSSION TITLE
PUT request gets stuck in endless loop

### DISCUSSION THREAD

**Discussion by ShevchenkoStar** [fastapi/fastapi#8695] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import logging

import uvicorn

from fastapi import FastAPI, Request
from fastapi.responses import JSONResponse
from fastapi.exceptions import RequestValidationError
from pydantic import BaseModel

app = FastAPI()


class Person(BaseModel):
    name: str
    age: int
    gender: str
    meta: dict


@app.put("/stucked_put")
def stuck_put(person: Person):
    print("never get here (1)")


@app.exception_handler(RequestValidationError)
async def validation_error_handler(request: Request, exc: RequestValidationError) -> JSONResponse:
    invalid_payload = await request.body()
    logging.warning(f"god {invalid_payload=}")
    print("never get here (2)")
    return JSONResponse(status_code=422, content={"errors": exc.errors()})


if __name__ == '__main__':
    uvicorn.run(
        "main:app",
        host="0.0.0.0",
        port=8080,
    )
```

### Description

**What we have:**
* simple fastAPi application
* PUT route with payload validation via pydantic
* custom RequestValidationError handler

When I try to send an invalid payload, it gets stuck in an endless loop.
request example
```
PUT http://localhost:8080/stucked_put
Content-Type: application/json

{}
```
**gets stuck** - I mean that request never is completed and I can't receive any response.
The request gets stuck on this line `invalid_payload = await request.body()`

The same behaviour with PATH, POST.

### Operating System

Linux

### Operating System Details

Ubuntu 20.04.3 LTS

### FastAPI Version

0.70.0

### Python Version

3.9.5

### Additional Context

* starlette==0.16.0
* pydantic==1.8.2
* uvicorn==0.15.0

**Comment by DavidBord** [comment]
How do you send the HTTP requests?
Can you please try with curl? 
For example:
`curl -XPUT  -H "Content-Type: application/json" http://localhost:8080/stucked_put --data '{}`

**Comment by ShevchenkoStar** [comment]
I tried in Pycharm, scratch.http file <[link](https://www.jetbrains.com/help/pycharm/http-client-in-product-code-editor.html#creating-http-request-files)>

curl does not work too
```
 curl -XPUT 'http://localhost:8080/stucked_put' -H 'Content-Type: application/json' --data '{}'
```

**Comment by Kludex** [comment]
The `request.body` stream was already consumed. This is normal.

**Comment by ShevchenkoStar** [comment]
@Kludex I understant that body is consumed but why body is not returned then.  
Python got stuck on a line of code `await request.body()` and never returns.
I thought that if body was consumed its content is saved in a attribute of request object BUT in debug mode I saw 
this line `if not hasattr(self, "_body"):` returns **True**. (**_body** attr. is not present in **self**).

**Comment by Kludex** [comment]
> @Kludex I understant that body is consumed but why body is not returned then.
Python got stuck on a line of code await request.body() and never returns.

As I said, the `request.body` stream was already consumed. You'll not be able to retrieve the body anymore, and that's why it keeps hanging.

You can get more context here: https://github.com/encode/starlette/issues/495. The exception handler is built using a middleware, jfyk.

**Comment by ShevchenkoStar** [comment]
@Kludex 
Thanks for link.
So this issue is related to the stalette. I guess we finish here?

**Comment by YuriiMotov** [comment, marked as answer]
This was fixed in [Starlette 0.28](https://www.starlette.io/release-notes/#0280-june-7-2023) and now works fine:
```
WARNING:root:god invalid_payload=b'{\n}'
never get here (2)
INFO:     127.0.0.1:39498 - "PUT /stucked_put HTTP/1.1" 422 Unprocessable Entity
```
